chain-signatures/node/src/indexer_eth.rs

```diff
@@ -428,13 +535,179 @@ pub async fn run(
+    if let Some(pos) = requests_indexed_write.iter().position(
+        |&BlockAndRequests {
+             block_number: num,
+             block_hash: _,
+             indexed_requests: _,
+         }| num >= block_number,
+    ) {
+        requests_indexed_write.insert(
+            pos,
+            BlockAndRequests::new(block_number, block_hash, indexed_requests.to_vec()),
+        );
+    } else {
+        requests_indexed_write.push_back(BlockAndRequests::new(
+            block_number,
+            block_hash,
+            indexed_requests.to_vec(),
+        ));
+    }
+}
+
+/// Polls for the latest finalized block and update the latest finalized blocks.
+/// Once finalized block gets updated, we fetch the blocks in the epoch ending with the finalized block.
+/// To ensure the blocks we fetched are proven, we fetch the latest finalized block from Helios.
+/// Then using an unstrusted RPC client, we go backwards to fetch the blocks in the epoch and check that the hash of the block's header == its next block's parent_hash.
+/// This is proven because the finalized block is already proven by helios, and by checking current block hash == next block's parent hash, we prove that each block is indeed included in the chain.
+async fn refresh_finalized_epoch(
+    helios_client: &Arc<EthereumClient<FileDB>>,
+    untrusted_rpc_client: &RootProvider<Http<AlloyClient>>,
+    finalized_block: &Arc<RwLock<u64>>,
+    finalized_epoch: &Arc<RwLock<HashMap<u64, alloy::primitives::B256>>>,
+) -> anyhow::Result<()> {
+    tracing::info!("Refreshing finalized epoch");
+    let Some(cur_finalized_block) = helios_client
+        .get_block_by_number(BlockTag::Finalized, false)
+        .await
+        .map_err(|err| anyhow!("Failed to fetch latest finalized block: {:?}", err))?
+    else {
+        return Err(anyhow!("Fetching finalized block return None"));
+    };
+
+    let last_finalized_block_number = {
+        let mut finalized_block_write = finalized_block.write().await;
+        if *finalized_block_write == 0 {
+            *finalized_block_write = cur_finalized_block.header.number;
+        }
+        *finalized_block_write
+    };
+
+    tracing::info!(
+        "Current finalized block number: {}, last finalized block number: {}",
+        cur_finalized_block.header.number,
+        last_finalized_block_number
+    );
+
+    if cur_finalized_block.header.number == last_finalized_block_number {
+        return Ok(());
+    }
+
+    {
+        *finalized_block.write().await = cur_finalized_block.header.number;
+    }
+
+    let latest_finalized_block_number = cur_finalized_block.header.number;
+
+    {
+        let mut epoch_write = finalized_epoch.write().await;
+        epoch_write.clear();
+        epoch_write.insert(
+            latest_finalized_block_number,
+            cur_finalized_block.header.inner.parent_hash,
+        );
```

Comment:
Is there a race condition here? Could it be that the task running `send_requests_when_final` never sees a finalized epoch because we overwrite it with the next one?

I think this might happen when we are catching up for some reason, or if the other task gets delayed somehow.

Reply:
The suggested architectural change to merge two tasks into one should get rid of this issue.